packages/liveblocks-react/src/factory.tsx

```diff
@@ -238,22 +235,64 @@ export function createRoomContext<
   TRoomEvent,
   TThreadMetadata
 > {
-  const RoomContext = React.createContext<Room<
-    TPresence,
-    TStorage,
-    TUserMeta,
-    TRoomEvent
-  > | null>(null);
+  type TRoom = Room<TPresence, TStorage, TUserMeta, TRoomEvent>;
+
+  const RoomContext = React.createContext<TRoom | null>(null);
+
+  const roomCache = new Map<string, { room: TRoom; leave: () => void }>();
+
+  /**
+   * A cached version of .enterRoom(), where we cache each room/leave pair on
+   * a per-RoomProvider instance basis.
+   */
+  function stableEnterRoom(
+    instanceId: string,
+    roomId: string,
+    options: EnterOptions<TPresence, TStorage>
+  ): { room: TRoom; leave: () => void } {
+    const key = `${instanceId}:${roomId}`;
+    const cached = roomCache.get(key);
+    if (cached) return cached;
+
+    const rv = client.enterRoom<TPresence, TStorage, TUserMeta, TRoomEvent>(
+      roomId,
+      options
+    );
 
-  function RoomProvider(props: RoomProviderProps<TPresence, TStorage>) {
+    // Wrap the leave function to also delete the cached value
+    const origLeave = rv.leave;
+    rv.leave = () => {
+      origLeave();
+      roomCache.delete(key);
+    };
+
+    roomCache.set(key, rv);
+    return rv;
+  }
+
+  function RoomProviderOuter(props: RoomProviderProps<TPresence, TStorage>) {
+    // This "Outer" hack mostly exists to make ID generation work well in
```

Comment:
Oof 😬

Reply:
Yep! I tried really hard doing it without it, but could not make it part of the same state/scope by definition, since all state will get blown away by the second render call in StrictMode. The only way to have a stable identifier is to keep that state outside of the component, unfortunately.

Reply:
I don't think I understand this correctly. From my understanding, using an external variable with `useRef` also results in a different id during second render.

```tsx
let lastInstanceId = 0;

function Test() {
  const instanceIdRef = useRef(`p${++lastInstanceId}`);

  console.log(lastInstanceId, instanceIdRef.current);

  return null;
```

This also logs:
```
1 p1
2 p2
```

Reply:
Correct! It's a little bit hard to explain, but I'll try.

At first I tried doing this:

```tsx
function Component() {
  const instanceIdRef = useRef(`p${++lastInstanceId}`);
  const [{room, leave}, /* setter */] = useState(() => stableEnterRoom(instaceIdRef, ...));
}
```

In StrictMode, among more things, "renders" (eg executing the functional component as a function) are performed twice, and `useState` hook initializer functions are called twice.

In this case, that means the following happens:

- 1st render of `Component()`:
  - `lastInstanceId` becomes `1`
  - `useRef` produces `"p1"`
  - `instanceIdRef` is `"p1"`
  - 1st execution of `useState` initializer calls: `stableEnterRoom("p1", ...)` -> result discarded
  - 2nd execution of `useState` initializer calls: `stableEnterRoom("p1", ...)` -> result returned as `{ room, leave }`
  - Let's call this `leave` function instance F1
  - **All local state and render result from this 1st render is now discarded**
- 2nd render of `Component()`:
  - `lastInstanceId` becomes `2`
  - `useRef` produces `"p2"`
  - `instanceIdRef` is `"p2"`
  - 1st execution of `useState` initializer calls: `stableEnterRoom("p2", ...)` -> result discarded
  - 2nd execution of `useState` initializer calls: `stableEnterRoom("p2", ...)` -> result returned as `{ room, leave }`
  - Let's call this `leave` function instance F2

At this point, we have a problem. The room is now "claimed" twice, and we have created two instances of `leave` unsubscribe callbacks (F1 and F2), but we haven't been able to "capture" the first instance (F1). Since we are not able to ever call F1, the room will never get freed, since it is never fully released. From the client's perspective, it's still in use.

---

Now let's move the generation of the ID to the outer component, and structure it like this:

```tsx
function Outer() {
  const instanceIdRef = useRef(`p${++lastInstanceId}`);
  return <Inner instanceIdRef={instanceIdRef} />;
}

function Inner({ instanceIdRef }) {
  const [{room, leave}, /* setter */] = useState(() => stableEnterRoom(instaceIdRef, ...));
}
```

- 1st render of `Outer()`:
  - `lastInstanceId` becomes `1`
  - `useRef` produces `"p1"`
  - Function produces JSX tree `<Inner instanceIfRef="p1" />`
  - **All local state and render result from this 1st render is now discarded**
- 2nd render of `Outer()`:
  - `lastInstanceId` becomes `2`
  - `useRef` produces `"p2"`
  - Function produces JSX tree `<Inner instanceIfRef="p2" />`

- 1st render of `Inner()` (called with "p2"):
  - 1st execution of `useState` initializer calls: `stableEnterRoom("p2", ...)` -> result discarded
  - 2nd execution of `useState` initializer calls: `stableEnterRoom("p2", ...)` -> result returned as `{ room, leave }`
  - Let's call this `leave` function instance F1
  - **All local state and render result from this 1st render is now discarded**
- 2nd render of `Inner()` (called with "p2"):
  - 1st execution of `useState` initializer calls: `stableEnterRoom("p2", ...)` -> result discarded
  - 2nd execution of `useState` initializer calls: `stableEnterRoom("p2", ...)` -> result returned as `{ room, leave }`
  - This is the same instance as F1

There now only exists exactly one instance of a `leave` function (F1). The key difference with the first approach is that the `Inner()` component will never get rendered with "p1" as the prop, and therefore this poses no issue in StrictMode.

In the second example's timeline, you can see that in total, 4 calls to `stableEnterRoom()` are being made, all with "p2" as the argument. The purpose of `stableEnterRoom()` is to cache the result of calling `enterRoom()` in the client so that when called for the same room ID (and by the same provider instance), that it will return the same, single `{ room, leave }` pair on every call. This is crucial in StrictMode, because if we did _not_ cache this result, it would create 4 `leave` functions/claims. This isn't problematic in and of itself of course, but… in StrictMode, we would not be able to capture 2 of those, because StrictMode would discard those results.

I realize this is complicated to explain, but hopefully this helps to understand and provides context!

Reply:
Thank you for the detailed explanation! Feels slightly hacky to do this but I understand the need to do this. Thank you for ensuring the changes are thorough! Very much appreciated ❤️

Reply:
I agree that it _is_ slightly hacky. One reason this must be done is because our `client.enter("my-room")` was pure before (it returned the same room however often you would call it with the same room ID argument). This is no longer the case. Our new `client.enterRoom("my-room")`, when called multiple times, will return the equivalent of this:

- `{ room: R1, leave: F1 }`
- `{ room: R1, leave: F2 }`
- `{ room: R1, leave: F3 }`
- `{ room: R1, leave: F4 }`

So even though the _room_ is stable, each call produces a new leave function (and thus the entire object returned is unstable). By caching these results, by doing that on a per-RoomProvider instance basis, we make the function pure again, which is what React expects.

Reply:
Writing this comment just made me realize a thing — we don't have to keep the state at the factory-closure level outside of React and keep cache keys as `p<instance-id>:<room-id>`! Instead, we can replace the generation of the instance ID ref and instead keep a cache there, local to the RoomProvider instance, with cache keys of just `<room-id>`. This way, it's just kept in local React state, solves the same problem, and makes leaking memory less of a concern, should there be a bug somewhere (because if the component is unmounted, React will clear the state that keeps the cache). Much nicer!

This is simplified in a2978a2d20297e9cbbb3066ba8609dd41f235c06 now!

Thanks for asking the right questions here!